### PR TITLE
provider/azure: added base ARM support in Azure provider.

### PR DESCRIPTION
**Pull request by aznashwan**
:exclamation: This PR was sent in to provide visibility to the common base from which work on Azure ARM-based resources can begin. It is in no way intended to be merged as-is.

This PR brings support for using the new ARM-based sdk in the Terraform Azure provider.
The respective authentication in the provider itself has been handled, as well as the renaming of old implementations and the addition of all the skeleton code for the new ARM-based implementation.

#### Notes on the naming conventions:

All changes follow the 'architecture' agreed upon in the issue opened on the ARM provider here: https://github.com/hashicorp/terraform/issues/3212
A quick rundown on the actual changes:

##### Old and new clients:

The new, larger `AzureClient` is now a composition of the smaller `asmClient` and `armClient`, each ones containing all the Clients for their respective API. Thus, the clients for both the APIs have been clearly separated and namespaced for clarity.

##### Old ASM-based resources which are currently not available in the ARM sdk:

Respective source files and containing functions have been renamed from "resource_azure_\*.go" to "resource_asm_\*.go" to maintain clarity and code order. Same applies for their test files.

They have **not** had the `use_asm_api` flag added to their schema as it would be pointless.

##### New ARM-based resources.

All the new resources which we can now define through the ARM api have had their skeleton code added to their respective "resource_arm_\*.go" file and now happily await implementation.

##### Resources definable through both APIs:

For the handful of resources which may be defined using both APIs, their ResourceFunc (and thus the Schemas they return) have had a `use_asm_api` flag added to their Schemas and have been left in their respective "resource_azure_\*.go" files all by themselves.

Their old ASM-based implementation has been moved into their respective "resource_asm_\*.go" files, with the same applying for their tests.

Each common resource has had a new "resource_arm_\*.go" file created for it and the skeletons of their CRUED functions placed there and awaiting implementation.

The Schema of these cross-API resources has been modified to use the implementation-switching functions for each resource, which, considering they are the same for each of them, have all been added in the file entitled "choice_functions.go".

#### Helping out:

As mentioned, skeleton code for all the [more urgently] desired resources has been added for utmost convenience. Please feel free to start implementing whichever one tickles your fancy the most.

:exclamation: to avoid duplicating work; do please sync up on IRC/leave notes here or on the open issue before starting working on a new resource implementation :exclamation:

### CODE REVIEW

**Comment by aznashwan** [comment]
Checklist of ARM resources to be implemented, (without tests):
- [x] resource group
- [x] virtual network
- [x] load balancer
- [ ] public IP
- [ ] subnet
- [ ] application gateway
- [ ] availability set
- [ ] dns server
- [ ] gateway
- [ ] gateway connection
- [ ] instance
- [ ] job
- [ ] job collection
- [ ] local network connection
- [ ] network interface
- [ ] security group
- [ ] security group rule
- [ ] storage service
- [ ] tag
- [ ] vm image
- [ ] vm extension image

File: builtin/providers/azure/provider.go [hunk 1]
```diff
@@ -39,108 +43,164 @@ func Provider() terraform.ResourceProvider {
 				Optional:    true,
 				DefaultFunc: schema.EnvDefaultFunc("AZURE_CERTIFICATE", ""),
 			},
+
+			// ARM-specific fields:
+			"arm_config_file": &schema.Schema{
```
Comment (jen20) [inline builtin/providers/azure/provider.go]: The current pattern (as of Terraform 0.6.7) here is to make the field accept the contents of the file, and then allow use of the `file` variable interpolation.

**Comment by xied75** [comment]
@aznashwan @jen20 

I got some free time to join the work here. It seems that on the IRC nobody is talking about anything at EU day time which is very inconvenient. So how many people is actually working on this? Is the coordinate really going on?

**Comment by jen20** [comment]
@xied75 I will be merging the current work in progress today once there are some minor updates made. After that we will split out remaining resources into separate issues under a milestone such that community contributors can pick them off one-by-one and there's some visibility over who is doing what - so please feel free to pick up any resources you want after that!

File: builtin/providers/azure/provider.go (hunk 1 above)
Comment (aznashwan) [inline builtin/providers/azure/provider.go]: @jen20: good catch; issue has been resolved.

**Comment by jen20** [comment]
Hi @aznashwan! After discussing with @phinze and doing some testing we decided this needs approaching slightly differently because of the validation requirements impact on usability when dealing with resources shared between ASM and ARM (for example, `resource_group_name` being required on ARM resources). I've started work on refactoring your stuff to match that model, and will link it here shortly. I'm still reasonably positive on having the initial support merged early this week though.

**Comment by jen20** [comment]
Hi @aznashwan - I've the process of merging this work into a new provider in #4226 - please feel free to comment on anything which concerns you there. For now I'm going to close this PR - if you have updates please target them at the `f-azure-arm` branch until that gets merged to master!

**Pull request closed without merging**